src/txn/txn.c

```diff
@@ -318,6 +319,75 @@ __wt_txn_bump_snapshot(WT_SESSION_IMPL *session)
     __txn_get_snapshot_int(session, false);
 }
 
+/*
+ * __wt_txn_snapshot_save_and_refresh --
+ *     Save the existing snapshot and allocate a new snapshot.
+ */
+int
+__wt_txn_snapshot_save_and_refresh(WT_SESSION_IMPL *session)
+{
+    WT_DECL_RET;
+    WT_TXN *txn;
+
+    txn = session->txn;
+
+    WT_RET(__wt_calloc_def(session, sizeof(WT_TXN_SNAPSHOT), &txn->backup_snapshot_data));
+
+    txn->backup_snapshot_data->snap_max = txn->snapshot_data.snap_max;
+    txn->backup_snapshot_data->snap_min = txn->snapshot_data.snap_min;
+    txn->backup_snapshot_data->snapshot_count = txn->snapshot_data.snapshot_count;
+
+    /* Take the backup of the snapshot only if the snapshot count is greater than 0. */
+    if (txn->snapshot_data.snapshot_count > 0) {
+        WT_ERR(__wt_calloc_def(session, sizeof(uint64_t) * S2C(session)->session_size,
+          &txn->backup_snapshot_data->snapshot));
+
+        /* Swap the snapshot pointers. */
+        __txn_swap_snapshot(&txn->snapshot_data.snapshot, &txn->backup_snapshot_data->snapshot);
+
+        /*
+         * __txn_get_snapshot_int will return without getting the new snapshot if the transaction
+         * already has a snapshot so clear the flag WT_TXN_HAS_SNAPSHOT.
+         */
+        F_CLR(txn, WT_TXN_HAS_SNAPSHOT);
+
+        /* Get the snapshot without publishing the shared ids. */
+        __wt_txn_bump_snapshot(session);
```

Comment:
I missed this in the previous review. I think we should bump the snapshot no matter the previous snapshot is 0 or not. If the previous snapshot has 0, we may bump to a snapshot with more snapshots.

Therefore, there is no need to pass in a bool here as well. If this function returns 0, we must have bumped the snapshot.

Reply:
Fixed in a5ecfa2de7d6c80534029e741e036ec37eb4e955